base/experimental.jl

```diff
@@ -152,6 +165,12 @@ macro compiler_options(args...)
                 a = a === false || a === :no  ? 0 :
                     a === true  || a === :yes ? 1 : error("invalid argument to \"infer\" option")
                 push!(opts.args, Expr(:meta, :infer, a))
+            elseif ex.args[1] === :max_methods
+                a = ex.args[2]
+                a = a === :default ? 3 :
+                  a isa Int ? (a > 0 ? a : error("\"max_methods\" must be positive.")) :
+                  error("invalid argument to \"max_methods\" option")
+                push!(opts.args, Expr(:meta, :max_methods, a))
```

Comment:
Does not enforce a max of 4

Reply:
I haven't actually tried values >4 (and I've only tried 4 in the sense that this was once the default in Julia).

@JeffBezanson said that inference time is potentially `O(N^N)` in max methods, so that anything >5 isn't viable.
I wrote the comment on the maximum with the intention of strongly warning against trying large numbers. E.g., someone on discourse wrote that they wanted 20, which Jeff said would never actually compile.

I should probably throw for larger values, in which case what is the best cutoff, 4 as the docstring currently says?
Or, we could just make a strong warning that increasing it is likely to cause super-linear inference regressions.

Reply:
Yeah, compilation with `max_methods > 4` is infeasible in general, so I think setting this default limitation seems reasonable.

Reply:
Okay, it now throws for `max_methods > 4`.